libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,825 @@
+        .note("skipping invalid events")
+        .primary(args_.x)
+        .primary(args_.res->source)
+        .emit(dh);
+      return false;
+    }
+    return true;
+  }
+
+  auto validate_y(const data& d, tenzir::location loc,
+                  diagnostic_handler& dh) const -> bool {
+    if (args_.ty == chart_type::line and is<caf::none_t>(d)) {
+      return true;
+    }
+    const auto ty = type::infer(d);
+    if (not ty) {
+      return false;
+    }
+    if (not ty->kind()
+              .is_any<int64_type, uint64_type, double_type, duration_type>()) {
+      diagnostic::warning("y-axis cannot have type `{}`", ty->kind())
+        .primary(std::move(loc))
+        .emit(dh);
+      return false;
+    }
+    if (args_.y_min or args_.y_max) {
+      const auto lty = args_.y_min ? type::infer(args_.y_min->inner)
+                                   : type::infer(args_.y_max->inner);
+      if (not lty or lty->kind() != ty->kind()) {
+        diagnostic::warning("limit has a different type `{}` from `y` type "
+                            "`{}`",
+                            lty->kind(), ty->kind())
+          .primary(args_.y_min ? args_.y_min->source : args_.y_max->source)
+          .note("skipping series")
+          .emit(dh);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Modified from `floor()`
+  auto floor(const series& xs) const -> series {
+    return match(
+      *xs.array,
+      [&](const arrow::DurationArray& array) -> series {
+        auto b
+          = duration_type::make_arrow_builder(arrow::default_memory_pool());
+        check(b->Reserve(array.length()));
+        for (auto i = int64_t{0}; i < array.length(); i++) {
+          if (array.IsNull(i)) {
+            check(b->AppendNull());
+            continue;
+          }
+          const auto val = array.Value(i);
+          const auto count = std::abs(args_.res->inner.count());
+          const auto rem = std::abs(val % count);
+          if (rem == 0) {
+            check(b->Append(val));
+            continue;
+          }
+          const auto floor = val >= 0 ? -rem : rem - count;
+          check(b->Append(val + floor));
+        }
+        return {duration_type{}, finish(*b)};
+      },
+      [&](const arrow::TimestampArray& array) -> series {
+        auto opts = make_round_temporal_options(args_.res->inner);
+        return {time_type{},
+                check(arrow::compute::FloorTemporal(array, std::move(opts)))
+                  .array_as<arrow::TimestampArray>()};
+      },
+      [&](const auto&) -> series {
+        TENZIR_UNREACHABLE();
+      });
+  }
+
+  friend auto inspect(auto& f, chart_operator2& x) -> bool {
+    return f.object(x)
+      .pretty_name("chart_operator2")
+      .fields(f.field("args_", x.args_));
+  }
+
+private:
+  chart_args args_;
+};
+
+template <chart_type Ty>
+class chart_plugin : public virtual operator_plugin2<chart_operator2> {
```

Comment:
This needs to be an `operator_parser_plugin`, and there has to be a separate `operator_inspection_plugin` for `chart_operator2`. Otherwise `chart_operator2` can't be serialized (you can check with `--dump-pipeline`, which goes through the serialization logic).